### PR TITLE
RavenDB-18201 - Sharding - Edit Document View Endpoints

### PR DESCRIPTION
**Pull request by lastav5**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18201

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                    var responseStream = RequestHandler.ResponseBodyStream();
+                    var count = stream.Read(buffer.Memory.Memory.Span); // can never wait, so no need for async
+                    while (count > 0)
+                    {
+                        await responseStream.WriteAsync(buffer.Memory.Memory.Slice(0, count), GetDataBaseShutDownToken());
+                        // we know that this can never wait, so no need to do async i/o here
+                        count = stream.Read(buffer.Memory.Memory.Span);
+                    }
+                }
+            }
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs]: `stream.CopyToAsync(RequestHandler.ResponseBodyStream(), HttpContext.AbortRequest)` ?

**Review by karmeli87 — APPROVED** [review]

File: src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+        protected AbstractAttachmentHandlerProcessorForGetAttachment([NotNull] TRequestHandler requestHandler, [NotNull] JsonContextPoolBase<TOperationContext> contextPool, Logger logger, bool isDocument) : base(requestHandler, contextPool)
+        {
+            Logger = logger;
+            _isDocument = isDocument;
+        }
+
+        protected abstract ValueTask<AttachmentResult> GetAttachmentAsync(TOperationContext context, string documentId, string name, AttachmentType type, string changeVector);
+
+        protected abstract CancellationToken GetDataBaseShutDownToken();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs]: Why not use RequestHandler.GetOperationToken?

File: src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                    HttpContext.Response.Headers["Content-Disposition"] = $"attachment; filename=\"{fileName}\"; filename*=UTF-8''{fileName}";
+                }
+                catch (ArgumentException e)
+                {
+                    if (Logger.IsInfoEnabled)
+                        Logger.Info($"Skip Content-Disposition header because of not valid file name: {attachment.Details.Name}", e);
+                }
+                try
+                {
+                    HttpContext.Response.Headers["Content-Type"] = attachment.Details.ContentType;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs]: can we move all of those to contants?

File: src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+                var attachment = await GetAttachmentAsync(context, documentId, name, type, changeVector);
+                if (attachment == null)
+                    return;
+
+                try
+                {
+                    var fileName = Path.GetFileName(attachment.Details.Name);
+                    fileName = Uri.EscapeDataString(fileName);
+                    HttpContext.Response.Headers["Content-Disposition"] = $"attachment; filename=\"{fileName}\"; filename*=UTF-8''{fileName}";
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs]: can we move all of those to contants?

File: src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+                    if (request.TryGet("Type", out string typeString) == false ||
+                        Enum.TryParse(typeString, out type) == false)
+                        throw new ArgumentException("The 'Type' field in the body request is mandatory");
+
+                    if (request.TryGet("ChangeVector", out changeVector) == false && changeVector != null)
+                        throw new ArgumentException("The 'ChangeVector' field in the body request is mandatory");
+                }
+
+                var attachment = await GetAttachmentAsync(context, documentId, name, type, changeVector);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs]: do we need to pass the token here as well and use it internally?

Comment (ppekrol) [reply src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs]: other places where we should use token?

Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs]: passed for sharded operation usage

File: src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs (hunk 4 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs]: done

File: src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs (hunk 3 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs]: done

File: src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs (hunk 2 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForGetAttachment.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Attachments/ShardedAttachmentHandlerProcessorForGetAttachment.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        public ShardedAttachmentHandlerProcessorForGetAttachment([NotNull] ShardedDatabaseRequestHandler requestHandler, bool isDocument) : base(requestHandler, requestHandler.ContextPool, requestHandler.Logger, isDocument)
+        {
+        }
+
+        protected override async ValueTask<AttachmentResult> GetAttachmentAsync(TransactionOperationContext context, string documentId, string name, AttachmentType type, string changeVector, CancellationToken token)
+        {
+            int shardNumber = RequestHandler.DatabaseContext.GetShardNumber(context, documentId);
+            var cmd = new GetAttachmentOperation.GetAttachmentCommand(context, documentId, name, type, changeVector);
+            var result = await RequestHandler.ShardExecutor.ExecuteSingleShardAsync(cmd, shardNumber, token);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Attachments/ShardedAttachmentHandlerProcessorForGetAttachment.cs]: we now can you ProxyCommand and simply the code?

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Attachments/ShardedAttachmentHandlerProcessorForPutAttachment.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        public ShardedAttachmentHandlerProcessorForPutAttachment([NotNull] ShardedDatabaseRequestHandler requestHandler) : base(requestHandler, requestHandler.ContextPool)
+        {
+        }
+
+        protected override async ValueTask<AttachmentDetails> PutAttachmentsAsync(TransactionOperationContext context, string id, string name, Stream requestBodyStream, string contentType, string changeVector, CancellationToken token)
+        {
+            int shardNumber = RequestHandler.DatabaseContext.GetShardNumber(context, id);
+            var op = new PutAttachmentOperation.PutAttachmentCommand(id, name, requestBodyStream, contentType, changeVector, validateStream: false);
+        
+            DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Stav, DevelopmentHelper.Severity.Normal,
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Attachments/ShardedAttachmentHandlerProcessorForPutAttachment.cs]: Would using ProxyCommand solve your problems?

File: src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForPutAttachment.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            using (ContextPool.AllocateOperationContext(out TOperationContext context))
+            using (var token = RequestHandler.CreateOperationToken())
+            {
+                var id = RequestHandler.GetQueryStringValueAndAssertIfSingleAndNotEmpty("id");
+                var name = RequestHandler.GetQueryStringValueAndAssertIfSingleAndNotEmpty("name");
+                var contentType = RequestHandler.GetStringQueryString("contentType", false) ?? "";
+                var requestBodyStream = RequestHandler.RequestBodyStream();
+                var changeVector = RequestHandler.GetStringFromHeaders("If-Match");
+
+                var result = await PutAttachmentsAsync(context, id, name, requestBodyStream, contentType, changeVector, token.Token);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForPutAttachment.cs]: Let's create `HandlePutAttachmentAsync` abstract method
1. In regular one also write a result
2. In sharded one we can use ProxyCommand

Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Attachments/AbstractAttachmentHandlerProcessorForPutAttachment.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Attachments/ShardedAttachmentHandlerProcessorForPutAttachment.cs (hunk 7 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Attachments/ShardedAttachmentHandlerProcessorForPutAttachment.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Attachments/ShardedAttachmentHandlerProcessorForGetAttachment.cs (hunk 6 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Attachments/ShardedAttachmentHandlerProcessorForGetAttachment.cs]: done

**Pull request merged**